Mathlib/Computability/PrimrecFilter.lean

```diff
@@ -0,0 +1,179 @@
+    filter_length_ne_zero_exists f L b
+  apply of_eq ?_ h
+  apply PrimrecPred.not (PrimrecRel.comp Primrec.eq (Primrec.comp list_length ?_) (const 0))
+  exact Primrec₂.comp (Primrec₂.swap (primrec₂.list_filter f hf)) snd fst
+
+/-- If f a b is decidable, then given L : List α and b : β, "g L b ↔ ∀ a L, f a b"
+is a primitive recurisve relation. -/
+lemma filter_forall (hf : PrimrecRel f) :
+    PrimrecRel fun (L : List α) ↦ fun b ↦ (∀ a ∈ L, f a b) := by
+  let g (b : β) := fun L ↦ List.filter (fun a ↦ f a b) L
+  have h (L : List α) (b : β) : (g b L).length = L.length ↔ (∀ a ∈ L, f a b) :=
+    filter_length_eq_length_for_all f L b
+  apply PrimrecRel.of_eq ?_ h
+  unfold PrimrecRel Primrec₂
+  rw [← PrimrecPred]
+  apply PrimrecRel.comp Primrec.eq (Primrec.comp list_length ?_) (Primrec.comp list_length fst)
+  exact Primrec₂.comp (Primrec₂.swap (primrec₂.list_filter f hf)) snd fst
+
+variable (f : ℕ → ℕ → Prop) [DecidableRel f]
+
+/-- If f a b is decidable, then for any fixed n and y,  "g n y ↔ ∃ x < n, f x y" is a
+primitive recursive relation. -/
+lemma bounded_exists (hf : PrimrecRel f) : PrimrecRel (fun n ↦ (fun y ↦ (∃ x < n, f x y))) := by
+  have h : PrimrecRel (fun n ↦ (fun y ↦ (∃ x ∈ range n, f x y))) :=
+    PrimrecRel.comp (filter_exists f hf) (Primrec.comp list_range fst) snd
+  exact PrimrecRel.of_eq h (by simp)
+
+/-- If f a b is decidable, then for any fixed n and y,  "g n y ↔ ∀ x < n, f x y" is a
+primitive recursive relation. -/
+lemma bounded_forall (hf : PrimrecRel f) : PrimrecRel (fun n ↦ (fun y ↦ (∀ x < n, f x y))) := by
+  have h : PrimrecRel (fun n ↦ (fun y ↦ (∀ x ∈ range n, f x y))) :=
+    PrimrecRel.comp (filter_forall f hf) (Primrec.comp list_range fst) snd
+  exact PrimrecRel.of_eq h (by simp)
+
+end PrimrecRel
+
+namespace Primrec
+
+/-- A helper lemma for proofs about bounded quantifiers on decidable relations. -/
+lemma nat_rel_list_filter (f : ℕ → ℕ → Prop) (s : ℕ) [DecidableRel f] (hf : PrimrecRel f) :
+    Primrec fun n ↦ ((List.range (s)).filter (fun y ↦ f y n)) := by
+  let g (n : ℕ) : ℕ → Option Nat := (fun y ↦ (if f y n = True then y else Option.none))
+  have h (n : ℕ) : (range (s)).filter (fun y ↦ f y n) = filterMap (g n) (List.range s) :=
+    primrec₂.filter_filtermap_lemma n f (range s)
+  simp only [h]
+  refine listFilterMap (Primrec.const (range s)) ?_
+  refine Primrec.ite ?_ (option_some_iff.mpr snd) (Primrec.const Option.none)
+  simp only [eq_iff_iff, iff_true]
+  exact PrimrecRel.comp hf snd fst
+
+end Primrec
+
+namespace PrimrecPred
+
+variable (f : ℕ → ℕ → Prop) (s : ℕ) [DecidableRel f]
+
+/-- If f a b is decidable, then for any fixed n and y, "∃ x < n, f x y" is a
+primitive recursive predicate in n. This is sometimes easier to work with than the fully
+general case involving a primitive recursive relation. -/
+lemma bounded_exists (hf : PrimrecRel f) : PrimrecPred (fun n ↦ ∃ y < s, (f y n)) := by
+  have h1 : (fun n ↦ decide (∃ y < s, f y n)) =
```

Comment:
You can make `have`s with parameters, like so:
```
have h (n) : decide (∃ y < s, f y n) = decide ((List.range s).filter (f · n) ≠ []) := by simp
```
(I have also used the trick with anonymous functions again: ` (f · n)` means the same as `fun y ↦ f y n`.)